arkouda_cws_hashing/server/lshMinMax.chpl

```diff
@@ -0,0 +1,129 @@
+module lshMinMax
+{
+
+  use CTypes;
+  use ServerConfig;
+  use MultiTypeSymEntry;
+  use ServerErrorStrings;
+  use MultiTypeSymbolTable;
+  use Message;
+  use Reflection;
+  use ServerErrors;
+  use Logging;
+
+
+  require '-lgsl','-lgslcblas';
+
+  extern {
+    #include "gsl/gsl_rng.h"
+    #include "gsl/gsl_randist.h"
+  }
+
+
+  // TODO: adapt 3D grid from 2.5D communication-avoiding matmul
+  // of Demmel and Solomonik to load-balance pair construction
+  // among LSH survivors analogous to Burkhardt's Toeplitz
+  // load-balancing strategy for all-pairs exact Jaccard similarity?
+  // Note that this was developed for dense matrices, and the
+  // LSH survivor output is likely to be sparse
+
+
+  proc cantorPairing(setEltIdx: int(64), hashIdx: int(64)): real(64) {
+
+      var sum: int(64) = setEltIdx + hashIdx;
+
+      var result: real(64) = (0.5 * sum * (sum + 1)) + hashIdx;
+
+      return result;
+  }
+
+
+  /* Returns a tuple consisting of aligned arrays of length numHashes*numSets containing 
+     sampled set elements and hashes */
+
+
+  proc getMinHashes(offsets: [?oD] int(64), setElts: [?sD] int(64),
+                    weights: [sD] real(64), numHashes: int(64)) throws {
+
+      var outD: domain(1) = {oD.first..numHashes*oD.last-1};
+
+
+/* TODO: define segments of these locally within the innermost loop, then assemble them into
+         the global output vectors later to avoid unnecessary communication over the network */
+
+      var preimages: [outD] int(64);
+      var minHashes: [outD] real(64);
+
+      /* CSR-style loop over a segmented array. Outer loop is data parallel. */
+
+      for offsetIdx in {offsets.domain.first..offsets.domain.last-1} {
+
+             var setStart: int(64) = offsets[offsetIdx];
+             var setFinal: int(64) = offsets[offsetIdx+1];
```

Comment:
Can change this.  In scientific computing CSR/CSC based datastructures are traditionally size n+1 (rather than n) with the final entry being the upper bound on the number of entries, since the implementations were typically with C arrays.  I'm certainly open to changing this (since I had to debug more than my usual share of "off-by-one" issues that were exacerbated by Chapel syntax, and it would be good to standardize things with segstring and segarray).

This actually brings to mind something that we used to do in MTGL and MTA-C (for the XMT) - manual Manhattan loop collapses for irregular prefix-sum operations.  I wonder if a similar idea would be applicable here.

Reply:
Can't remember if Manhattan loop collapses were appropriate only for intranode parallelism, or weird architectures like the MTA that offered equal-cost distributed memory access (and can't find my usual reference).  But maybe the zippered iteration in the outer forall loop is not completely collapsing the iterations over set elements, since there is an inner (serial) loop?  And how much do you all worry about parallelizing the inner loop over set elements in a segarray or segstring?  I'm guessing this isn't much of a problem for text data, but could be a massive problem for skewed "power-law" graphs with O(n) neighbor sets (which admittedly isn't the necessarily the application LSH techniques like mine are used for, even though it is a legitimate use case)